webapp/application.py

```diff
@@ -125,4 +131,178 @@ def application_page(token):
+        or not candidate["first_name"].lower() == name
+    ):
+        flask.abort(404)
+
+    candidate_name = candidate["first_name"]
+    position = application["jobs"][0]["name"]
+    hiring_lead = hiring_lead["name"]
+
+    # Sanitize and parse user input
+    email = parseaddr(flask.request.form["email"])[1]
+    candidate_email = parseaddr(candidate["email_addresses"][0]["value"])[1]
+
+    if flask.request.form["withdrawal-reason-other"]:
+        withdrawal_reason = flask.request.form["withdrawal-reason-other"]
+    else:
+        withdrawal_reason = flask.request.form["withdrawal-reason"]
+
+    # Reject if user typed the wrong email
+    if not candidate_email == email:
+        flask.abort(404)
+
+    send_mail(
+        to_email=[candidate_email],
+        subject="Withdraw Application Confirmation",
+        message=flask.render_template(
+            "applications/_activate-email.html",
+            applicant_name=candidate_name,
+            position=position,
+            hiring_lead=hiring_lead,
+            verification_link=confirmation_token(
+                candidate_email,
+                withdrawal_reason,
+                candidate_id,
+                application_id,
+                hiring_lead,
+            ),
+        ),
+    )
+    return redirect(flask.request.referrer + "#withdrawal-requested")
+
+
+def confirmation_token(
+    email, withdrawal_reason, candidate_id, application_id, hiring_lead
+):
+    payload = {
+        "email": email,
+        "withdrawal_reason": withdrawal_reason,
+        "candidate_id": candidate_id,
+        "application_id": application_id,
+        "hiring_lead": hiring_lead,
+    }
+    token = json.dumps(payload)
+    return verification_token_cipher.encrypt(token)
+
+
+# Send email if configured
+smtp_server = os.getenv("SMTP_SERVER")
+smtp_user = os.getenv("SMTP_USER")
+smtp_pass = os.getenv("SMTP_PASS")
+smtp_sender_address = os.getenv("SMTP_SENDER_ADDRESS")
+
+
+def send_mail(
+    to_email,
+    subject,
+    message,
+):
+    # import smtplib
+    try:
+        if smtp_server:
+            msg = EmailMessage()
+            msg["Subject"] = subject
+            msg["From"] = smtp_sender_address
+            msg["To"] = ", ".join(to_email)
+            msg.set_content(message, subtype="html")
+
+            server = SMTP(smtp_server)
+            if smtp_user and smtp_pass:
+                server.ehlo()
+                server.starttls()
+                server.ehlo()
+                server.login(smtp_user, smtp_pass)
+            server.send_message(msg)
+            server.quit()
+            print("successfully sent the email")
+        else:
+            print(message)
+    except Exception as e:
+        print(f"Error: unable to send email: {e}")
+
+
+def _post_api(url, headers, payload={}):
+    return session.post(
+        url,
+        auth=(harvest, ""),
+        data=json.dumps(payload),
+        headers=headers,
     )
+
+
+def get_reason_id(withdrawal_reason):
```

Comment:
I'm assuming we're doing this because we need to send this ID to Harvest, but Greenhouse doesn't provide any way to get the ID. Is that right? If so, this might be the best we can do, but it's a shame because this solution is pretty brittle. I doubt they consider those exact strings to be part of their API contract, so they could easily change in which case these IDs would no longer match up.

Reply:
When rejecting an application you may want (optionally) to provide the rejection reason, and that's done passing the id, that's right. But this function seems to be mapping the option selected in the front-end with the id, so this string could be anything as long as it matches the `value` attribute in the form. It doesn't need to be the specific reason as defined in GH. A `dict` mapping rejection reason with id for easy lookup would do the trick, IMHO.

Reply:
Sorry, I didn't quite follow that. Are you suggesting this function should stay the same? Or change?

Reply:
Sorry, I just meant that these strings were defined [here](https://github.com/canonical/canonical.com/pull/593/files#diff-3ec962bc637c818d3d11084bdbbd92ef5e84c768ef44910b13040681554435fcR14), so it won't break if somebody renames it in GH. About the function, it can stay the same, but I suggested that having a dict may be easier than an if statement. @nottrobin 
```py
rejection_reasons_ids = {
   "reason 1": 1111,
   "reason 2": 2222
   ...
}
```

Reply:
Sorry so these strings come through from our own form? If that's the case then we definitely shouldn't be using whole sentences like this as our `name` (basically the ID). If we do that, some future developer is clearly going to think those values can just be reworded without impact.

I suppose maybe this should be a dictionary of IDs to descriptions that we define at a more global level and pass through to the template? That would be much safer.